src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLHTMLGetFeatureOutputFormat.java

```diff
@@ -0,0 +1,195 @@
+        SimpleFeatureCollection fc = (SimpleFeatureCollection) featureCollection;
+        ReferencedEnvelope projectedBbox = extractBbox(fc);
+        LayerInfo layerInfo = gs.getCatalog().getLayerByName(fc.getSchema().getTypeName());
+        CoordinateReferenceSystem crs = projectedBbox.getCoordinateReferenceSystem();
+        ProjType projType = parseProjType(request);
+        double longitude;
+        double latitude;
+        ReferencedEnvelope geographicBox;
+        try {
+            geographicBox = projectedBbox.transform(DefaultGeographicCRS.WGS84, true);
+            longitude = geographicBox.centre().getX();
+            latitude = geographicBox.centre().getY();
+        } catch (TransformException | FactoryException e) {
+            throw new ServiceException("Unable to transform bbox to WGS84", e);
+        }
+
+        boolean flipCoords = CRS.getAxisOrder(crs).equals(CRS.AxisOrder.NORTH_EAST);
+        double lat = flipCoords ? longitude : latitude;
+        double lon = flipCoords ? latitude : longitude;
+
+        MapMLHTMLOutput output =
+                new MapMLHTMLOutput.HTMLOutputBuilder()
+                        .setLongitude(lon)
+                        .setLatitude(lat)
+                        .setRequest(httpRequest)
+                        .setLayerLabel(layerInfo.getTitle())
+                        .setProjType(projType)
+                        .setProjectedBbox(projectedBbox)
+                        .setSourceUrL(buildGetFeature(request))
+                        .build();
+
+        // write to output
+        OutputStreamWriter osw = new OutputStreamWriter(out, gs.getSettings().getCharset());
+        osw.write(output.toHTML());
+        osw.flush();
+    }
+
+    private ProjType parseProjType(Request request) throws ServiceException {
+        try {
+            Map<String, Object> rawKvp = request.getRawKvp();
+            String srs;
+            if (rawKvp.containsKey("SRSNAME")) {
+                srs = (String) rawKvp.get("SRSNAME");
+            } else if (rawKvp.containsKey("SRS")) {
+                srs = (String) rawKvp.get("SRS");
+            } else {
+                srs = "EPSG:4326";
+            }
+            return ProjType.fromValue(srs.toUpperCase());
+        } catch (IllegalArgumentException | FactoryException iae) {
+            // figure out the parameter name (version dependent) and the actual original
+            // string value for the srs/crs parameter
+            String parameterName =
+                    Optional.ofNullable(request.getVersion())
+                            .filter(v -> v.equals("1.0.0"))
+                            .map(v -> "srs")
+                            .orElse("srsName");
+            Map<String, Object> rawKvp = Dispatcher.REQUEST.get().getRawKvp();
+            String value = (String) rawKvp.get("SRSNAME");
+            if (value == null) value = (String) rawKvp.get("SRS");
+            throw new ServiceException(
```

Comment:
Here, we are using the same logic that is applied by WMS when producing the HTML. Only the 4 MapML supported CRSs  will be accepted.

Reply:
Without having time to get into the code, I must point out that we currently can serve any WFS SRS into MapML, realizing that we can't view that content without a custom TCRS definition, but it still may be useful to serve it for completeness' sake.  The technique I've used in the viewer is to disallow / throw on `projection` values that contain a `:` colon.  So when serving a text/mapml document as EPSG:1234 for example, we could create a custom TCRS definition for `EPSG1234` and serve that.  If the projection is unknown i.e. no custom TCRS definition exists, the viewer reports a different problem. The purpose of this is to avoid having people think that EPSG:codes code be uses as TCRS names directly.  i.e. you need the extra defintional stuff provided by a TCRS, it's not the same thing as an EPSG:code.